### PR TITLE
Connect frontend lobby to socket.io

### LINKED ISSUE #50: Integrate lobby sockets into frontend pages
Emits the following client initiated events:

- [x] connect
- [x] settings:update

Listens for the following server initiated events:

- [x] navigate
- [x] host
- [x] players:initial
- [x] players:add
- [x] players:remove
- [x] settings:initial
- [x] settings:update

For a detailed explanation of these events see our [Socket.IO Events Wiki](https://github.com/SE750-DART/nerds-oppose-society/wiki/Socket.IO-Events)

To get started implementing these events check out the Socket.IO docs on how to [connect a client](https://socket.io/docs/v4/client-initialization/) and [emitting events](https://socket.io/docs/v4/emitting-events/).

I also recommend looking at [this tutorial](https://socket.io/get-started/private-messaging-part-1/) from Socket.IO as it gives more practical examples than their docs. Of particular note are the following:

- Client initialisation will want to be done in a file that exports the socket object to be reused where needed. , and `autoConnect: false` as we only want to connect once the user has created a player: 
Screen Shot 2021-05-05 at 8 25 24 AM
- Additionally we will need to use `auth: { gameCode: "42069", playerId: "asger0uteu" }` in order to authenticate with the backend when we finally connect:
Screen Shot 2021-05-05 at 8 25 37 AM

### PR DESCRIPTION
**Pull request by RawiriHohepa**
Closes #50 

Emits the following client initiated events:

- [x] connect
- [x] settings:update

Listens for the following server initiated events:

- [x] navigate
- [x] host
- [x] players:initial
- [x] players:add
- [x] players:remove
- [x] settings:initial
- [x] settings:update

For a detailed explanation of these events see our [Socket.IO Events Wiki](https://github.com/SE750-DART/nerds-oppose-society/wiki/Socket.IO-Events)

### CODE REVIEW

File: frontend/src/ContextProviders/PlayersContextProvider.tsx
Comment (taitfuller) [inline frontend/src/ContextProviders/PlayersContextProvider.tsx:13]: We should similarly add a `me` field to this

File: frontend/src/GameRouter.tsx
Comment (taitfuller) [inline frontend/src/GameRouter.tsx:41]: To reduce network traffic only the setting being updated is sent to the client. I recommend using something similar to the following:

```
socket.on("settings:update", (setting, value) => {
    let key;
    switch (setting) {
      case "MAX_PLAYERS":
        key = "maxPlayers";
        break;
      case "ROUND_LIMIT":
        key = "roundLimit";
        break;
      default:
        key = null;
    }
    if (key !== null) {
      const newSettings = {
        ...settings,
        [key]: value,
      };
      setSettings(newSettings);
    }
  });
```

Comment (taitfuller) [reply frontend/src/GameRouter.tsx:41]: I could also look into having the server return `setting` as the appropriate key to remove this case if you prefer

File: frontend/src/pages/LobbyPage/index.tsx
Comment (taitfuller) [inline frontend/src/pages/LobbyPage/index.tsx:26]: Love some 🖨🍝 (`setMaxPlayers()`)

File: frontend/src/pages/LobbyPage/index.tsx
Comment (taitfuller) [inline frontend/src/pages/LobbyPage/index.tsx:55]: `newRoundLimit`

File: frontend/src/pages/LobbyPage/index.tsx
Comment (taitfuller) [inline frontend/src/pages/LobbyPage/index.tsx:17]: Should add `const { me, host, players } = useContext(PlayersContext)` to this component to disable setting input for non-host and display the number of players in the `<h2>`

File: frontend/src/pages/LobbyPage/index.tsx
Comment (taitfuller) [inline frontend/src/pages/LobbyPage/index.tsx:83]: Add `disabled={me !== host}`

File: frontend/src/pages/NicknamePage/index.tsx
Comment (taitfuller) [inline frontend/src/pages/NicknamePage/index.tsx:50]: Add `setMe(nickname)`

File: frontend/src/pages/NicknamePage/index.tsx
Comment (taitfuller) [inline frontend/src/pages/NicknamePage/index.tsx:17]: Should add `const { setMe } = useContext(PlayersContext) to this component`

File: backend/src/models/index.ts
Comment (taitfuller) [inline backend/src/models/index.ts:7]: Will be good to refactor these into a common library in the future 👍🏼

**Review by taitfuller — COMMENTED** [review]
Amazing work man!

Looks like you've made a few 🖨🍝 mistakes here & there. The biggest change required is how the settings are updated since the server only send the setting that has changed to the client - have have detailed a potential fix in my comment.

Would be great if you could add a `me` attribute to the `PlayerContext` and use this context in the `LobbyPage` and `PlayerList` component!

File: frontend/src/pages/LobbyPage/index.tsx
Comment (taitfuller) [reply frontend/src/pages/LobbyPage/index.tsx:17]: This is no longer required as of #68 as we can use the players `id` persisted with `use-persisted-state` matched against other player ids in the `PlayersContext`

File: frontend/src/ContextProviders/PlayersContextProvider.tsx
Comment (taitfuller) [reply frontend/src/ContextProviders/PlayersContextProvider.tsx:13]: This is no longer required as of #68 as we can use the players `id` persisted with `use-persisted-state` matched against other player ids in the `PlayersContext`

**Comment by taitfuller** [comment]
I recommend holding off merging this until #65 & #68 have been merged 👍🏼

File: frontend/src/pages/NicknamePage/index.tsx
Comment (taitfuller) [inline frontend/src/pages/NicknamePage/index.tsx:25]: I'm encountering the following error on the NicknamePage component:

Screen Shot 2021-05-07 at 3 12 24 AM

It appears that this webpage offers a potential fix by wrapping the call for this `async` with a check for if the component is mounted

File: frontend/src/GameRouter.tsx
Comment (taitfuller) [inline frontend/src/GameRouter.tsx:63]: We need to bind these listeners inside a useEffect(), alongside unbinding when the component unmounts. An example of this can be seen here: [https://dev.to/bravemaster619/how-to-use-socket-io-client-correctly-in-react-app-o65](https://dev.to/bravemaster619/how-to-use-socket-io-client-correctly-in-react-app-o65)

File: frontend/src/pages/NicknamePage/index.tsx
Comment (taitfuller) [inline frontend/src/pages/NicknamePage/index.tsx:39]: I wonder if this should be moved to GameRouter to be included in the useEffect() & be unmounted 🤔

File: frontend/src/pages/NicknamePage/index.tsx
Comment (taitfuller) [inline frontend/src/pages/NicknamePage/index.tsx:45]: Should `playerId` not be a dependency of the hook?

File: frontend/src/GameRouter.tsx
Comment (taitfuller) [reply frontend/src/GameRouter.tsx:63]: Similarly it seems that `useCallback()` should be used for performance improvements too

**Review by taitfuller — CHANGES_REQUESTED** [review]
Overall great functionality man! So cool to see it all coming together 🙌🏼

There is pretty serious performance issue in `GameRouter` caused by the Socket.io event listeners not being unmounted eventually causing clients to freeze after a number of server events modifying state as re-render binds additional listeners without unbinding old ones. [This webpage](https://dev.to/bravemaster619/how-to-use-socket-io-client-correctly-in-react-app-o65) seems to be a good reference on how to achieve this, while [this one] mentions calling `socket.disconnect()` which we should remember to call when the player ends a game and returns to the Home Screen.

Related to this note, I do wonder if we should reconsider moving the Players and Punchlines Contexts within the GameRouter component to not cause excessive mounting-unmounting cycles for Socket.io listeners and potentially missing server issues events.

One more thing, it appears the NicknamePage has a small memory leak caused by modifying state inside an `async` function within the `useEffect()` after the component has been unmounted. [This webpage](https://www.debuggr.io/react-update-unmounted-component/) details a potential fix for this.

File: frontend/src/pages/NicknamePage/index.tsx
Comment (taitfuller) [reply frontend/src/pages/NicknamePage/index.tsx]: This is no longer required as all player IDs are returned and the current player has their playerId stored in persistent state

File: frontend/src/pages/NicknamePage/index.tsx
Comment (taitfuller) [reply frontend/src/pages/NicknamePage/index.tsx]: No longer required

**Comment by RawiriHohepa** [comment]
> 
> Related to this note, I do wonder if we should reconsider moving the Players and Punchlines Contexts within the GameRouter component to not cause excessive mounting-unmounting cycles for Socket.io listeners and potentially missing server issues events.
> 

@taitfuller this won't work since the context is used within GameRouter, so the providers have to be at least one level above GameRouter

File: frontend/src/GameRouter.tsx
Comment (taitfuller) [inline frontend/src/GameRouter.tsx:88]: I'd add `[socket]` as a dependency to this hook 👍🏼 Tested it out locally & seems to work sweet

**Review by taitfuller — COMMENTED** [review]
Looks good man! Only thing I'd suggest is if `[socket]` should be added as a dependency to the `useEffect()` in `GameRouter`.

Another small think I've noticed is occasionally the persisted `playerId` & `token` get set to `""` & the player no longer has a record of who they are! Perhaps we should look at adding `playerId` to the context if this can't be quickly resolved.

**Review by taitfuller — APPROVED** [review]
Let's go!!! 📈🚀🌝

**Pull request merged**
